### PR TITLE
include splitting of the carbon budget and calculate the cumulative cost for the myopic approach

### PR DESCRIPTION
**Pull request by martavp**
This pull request:

- changes column name in powerplants database from "YearCommissioned" to "DateIn", solving[ issue 85](https://github.com/PyPSA/pypsa-eur-sec/issues/85)

- proposes a quick fix for [issue 79 ](https://github.com/PyPSA/pypsa-eur-sec/issues/79)

- For the myopic option, given a carbon budget, distributes it across the planning_horizons following an exponential or beta decay. 

A file with CO2 limit in every planning horizon and a plot showing historical and planned CO2 emissions are saved in the results
**This pr requires updating to UNFCCC_v23.csv in the data bundle** (because emissions in 2018, which are assumed to remain constant in 2019  and 2020, and subtracted from the carbon budget)

- Creates an additional file in results/csvs including the cumulative costs with different social discount rate. 
We could also include the social discount rate in the config file, but I think keeping the results with various discount rates forces the user to think about the implications of this assumption.

- Allows specifying an option to alter the capital cost of carriers by a factor indicated in the config file, eg: solar+c0.5 or specifying an option to alter the maximum capacity to be installed, eg: solar+p3 implies assuming 3 times the reference potential

### CODE REVIEW

File: config.default.yaml [hunk 1]
```diff
@@ -24,11 +24,16 @@ scenario:
   # B for biomass supply, I for industry, shipping and aviation
   # solarx or onwindx changes the available installable potential by factor x
   # dist{n} includes distribution grids with investment cost of n times cost in data/costs.csv
+  # solar+c0.5 reduces the capital cost of solar to 50\% of reference value
```
Comment (nworbmot) [inline config.default.yaml:27]: Is the intention that when you do solar+p0.5 it reduces the potential p_nom_max by 50%? This is like solar0p5, right?

File: Snakefile [hunk 2]
```diff
@@ -8,7 +8,7 @@ wildcard_constraints:
     clusters="[0-9]+m?",
     sectors="[+a-zA-Z0-9]+",
     opts="[-+a-zA-Z0-9]*",
-    sector_opts="[-+a-zA-Z0-9]*"
+    sector_opts="[-+a-zA-Z0-9\.\s]*"
```
Comment (nworbmot) [inline Snakefile:11]: Excellent, this allows us to avoid problems with solar0.5 etc.

File: config.default.yaml [hunk 3]
```diff
@@ -24,11 +24,16 @@ scenario:
...
+  # for myopic/perfect foresight cb states the carbon budget in GtCO2 (cumulative
```
Comment (nworbmot) [inline config.default.yaml:28]: Maybe include example at beginning of line, e.g. cb40ex0?

File: scripts/build_biomass_potentials.py [hunk 4]
```diff
@@ -57,7 +58,12 @@ def build_biomass_potentials():
...
         snakemake.output = Dict()
         snakemake.output['biomass_potentials'] = 'data/biomass_potentials.csv'
+        snakemake.output['biomass_potentials_all']='resources/biomass_potentials_all.csv'
         with open('config.yaml', encoding='utf8') as f:
             snakemake.config = yaml.safe_load(f)
-
+    
+    if 'Secondary Forestry residues sawdust' in snakemake.config['biomass']['classes']['solid biomass']:
+        snakemake.config['biomass']['classes']['solid biomass'].remove('Secondary Forestry residues sawdust')
+        snakemake.config['biomass']['classes']['solid biomass'].append('Secondary Forestry residues – sawdust')
```
Comment (nworbmot) [inline scripts/build_biomass_potentials.py:67]: Hmm, it's a bit hacky. Maybe add a comment that it is a hack, to be replaced once snakemake is unicode-conform.

**Comment by nworbmot** [comment]
Hi @martavp thanks for implementing these features! Looks mostly good to me, see comments above. Particularly the plotting should be separated or called in a separate function. Perhaps the script can output a CSV, which is then plotted in a separate script? It would be good to unify the ways of modifying p_nom_max and the capital cost.

File: config.default.yaml (hunk 1 above)
Comment (nworbmot) [reply config.default.yaml:27]: We could replace solarx and onwindx for altering the installable potentials with your notation.

**Comment by nworbmot** [comment]
It would in general be better, if possible, to split these different changes into different pull requests, although I know this is difficult and creates lots of extra branches, but it's easier to review and comprehend afterwards.

File: scripts/prepare_sector_network.py [hunk 5]
```diff
@@ -45,6 +47,206 @@
...
+                     markeredgecolor='black', label='EU commited target')
+            
+    ax1.legend(fancybox=True, fontsize=18, loc=(0.01,0.01), 
+                       facecolor='white', frameon=True) 
+            
+    path_cb_plot = snakemake.config['results_dir'] + snakemake.config['run'] + '/graphs/'             
+    if not os.path.exists(path_cb_plot):
+        os.makedirs(path_cb_plot)
+    print('carbon budget distribution saved to ' + path_cb_plot + 'carbon_budget_plot.pdf')
+    plt.savefig(path_cb_plot+'carbon_budget_plot.pdf', dpi=300)
```
[the remark this answers is not in the document]
Comment (martavp) [reply scripts/prepare_sector_network.py:248]: The reason why I did not include this as a snakemake output is that prepare_sector_network.py is shared between the overnight and myopic options but only the myopic produces a carbon_budget_distribution file (csv and pdf)

**Comment by martavp** [comment]
Thanks, @nworbmot. I mostly agree with your comments. I will take care of implemented the requested changes in the following days.

File: config.default.yaml (hunk 1 above)
Comment (martavp) [reply config.default.yaml:27]: Done here https://github.com/PyPSA/pypsa-eur-sec/pull/90/commits/fd6239db5c6e5f0a711f8fa99e9048b0d023f3b4

File: scripts/prepare_sector_network.py [hunk 6]
```diff
@@ -45,6 +47,206 @@
...
+        m=(1+np.sqrt(1+r*T))/T
+        CO2_CAP[o] = [(e_0/e_1990)*(1+(m+r)*(t-t_0))*np.exp(-m*(t-t_0)) for t in planning_horizons]
+                
+    CO2_CAP.to_csv(path_cb + 'carbon_budget_distribution.csv', sep=',', 
+                   line_terminator='\n',  float_format='%.3f') 
+ 
+    """
+    Plot historical carbon emissions in the EU and decarbonization path
+    """ 
+    import matplotlib.pyplot as plt
```
[the remark this answers is not in the document]
Comment (martavp) [reply scripts/prepare_sector_network.py:190]: Plotting has been moved to plot_summary.py https://github.com/PyPSA/pypsa-eur-sec/pull/89/commits/e180931ce1aeed2df8a340836cda75ff35f00029

File: config.default.yaml (hunk 3 above)
Comment (martavp) [reply config.default.yaml:28]: Yes, https://github.com/PyPSA/pypsa-eur-sec/pull/89/commits/faebbc493b78b6ec386e19bae2b1b43ef83e0c2a

File: scripts/build_biomass_potentials.py (hunk 4 above)
Comment (martavp) [reply scripts/build_biomass_potentials.py:67]: "hackyness" acknowledged https://github.com/PyPSA/pypsa-eur-sec/pull/88/commits/918d803c0d9b56f7f67898c93b7b7a29a9a88401

**Comment by martavp** [comment]
This pull request has been split in three

https://github.com/PyPSA/pypsa-eur-sec/pull/88

https://github.com/PyPSA/pypsa-eur-sec/pull/89

https://github.com/PyPSA/pypsa-eur-sec/pull/90

**Comment by nworbmot** [comment]
Thanks Marta! I didn't mean to cause you extra work to split it up this time, only next time, but thanks it's very helpful!

File: scripts/prepare_sector_network.py (hunk 6 above)
[the remark this answers is not in the document]
Comment (nworbmot) [reply scripts/prepare_sector_network.py:190]: Great, thanks Marta!

File: scripts/build_biomass_potentials.py (hunk 4 above)
Comment (nworbmot) [reply scripts/build_biomass_potentials.py:67]: ;-)

File: config.default.yaml (hunk 3 above)
Comment (nworbmot) [reply config.default.yaml:28]: Great, thanks!

**Comment by martavp** [comment]
No worries. I prefer to keep it tidy already this time.

**Pull request closed without merging**